readthedocs/core/unresolver.py

```diff
@@ -1,99 +1,284 @@
+        version = parent_project.versions.filter(
+            slug=parent_project.default_version
+        ).first()
+        if version:
+            return parent_project, version, file
+        return parent_project, None, None
 
-        # Handle our backend storage not supporting directory indexes,
-        # so we need to append index.html when appropriate.
-        if not filename or filename.endswith('/'):
-            # We need to add the index.html to find this actual file
-            filename += 'index.html'
-
-        log.debug(
-            'Unresolver parsed.',
-            project_slug=final_project.slug,
-            lang_slug=lang_slug,
-            version_slug=version_slug,
-            filename=filename,
+    def _unresolve_path(self, parent_project, path, check_subprojects=True):
+        """
+        Unresolve `path` with `parent_project` as base.
+
+        If the returned project is `None`, then we weren't able to
+        unresolve the path into a project.
+
+        If the returned version is `None`, then we weren't able to
+        unresolve the path into a valid version of the project.
+
+        :param parent_project: The project that owns the path.
+        :param path: The path to unresolve.
+        :param check_subprojects: If we should check for subprojects,
+         this is used to call this function recursively.
+
+        :returns: A tuple with: project, version, and file name.
+        """
+        # Multiversion project.
+        if not parent_project.single_version:
+            response = self._match_multiversion_project(
+                parent_project=parent_project,
+                path=path,
+            )
+            if response:
+                return response
+
+        # Subprojects.
+        if check_subprojects:
+            response = self._match_subproject(
+                parent_project=parent_project,
+                path=path,
+            )
+            if response:
+                return response
+
+        # Single version project.
+        if parent_project.single_version:
+            response = self._match_single_version_project(
+                parent_project=parent_project,
+                path=path,
+            )
+            if response:
+                return response
+
+        return None, None, None
+
+    @staticmethod
+    def get_domain_from_host(host):
+        """
+        Get the normalized domain from a hostname.
+
+        A hostname can include the port.
+        """
+        return host.lower().split(":")[0]
+
+    # TODO: make this a private method once
+    # proxito uses the unresolve method directly.
+    def unresolve_domain(self, domain):
+        """
+        Unresolve domain by extracting relevant information from it.
+
+        :param str domain: Domain to extract the information from.
+        :returns: A tuple with: the project slug, domain object, and if the domain
+         is from an external version.
```

Comment:
It would be good to make this a dataclass or similar. Returning these structured tuples is always hard to parse mentally and keep that in mind while working with this code and switching between files